src/Microsoft.Azure.ServiceBus/MessageReceivePump.cs

```diff
@@ -0,0 +1,266 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.Azure.ServiceBus
+{
+    using System;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.ServiceBus.Primitives;
+
+    public sealed class MessageReceivePump
+    {
+        const int MaxInitialReceiveRetryCount = 3;
+        static readonly TimeSpan ServerBusyExceptionBackoffAmount = TimeSpan.FromSeconds(10);
+        static readonly TimeSpan OtherExceptionBackoffAmount = TimeSpan.FromSeconds(1);
+        readonly Func<BrokeredMessage, CancellationToken, Task> onMessageCallback;
+        readonly OnMessageOptions onMessageOptions;
+        readonly MessageReceiver messageReceiver;
+        readonly CancellationToken pumpCancellationToken;
+        readonly SemaphoreSlim maxConcurrentCallsSemaphoreSlim;
+
+        public MessageReceivePump(
+            MessageReceiver messageReceiver,
+            OnMessageOptions onMessageOptions,
+            Func<BrokeredMessage, CancellationToken, Task> callback,
+            CancellationToken pumpCancellationToken)
+        {
+            if (messageReceiver == null)
+            {
+                throw new ArgumentNullException(nameof(messageReceiver));
+            }
+
+            this.messageReceiver = messageReceiver;
+            this.onMessageOptions = onMessageOptions;
+            this.onMessageCallback = callback;
+            this.pumpCancellationToken = pumpCancellationToken;
+            this.maxConcurrentCallsSemaphoreSlim = new SemaphoreSlim(this.onMessageOptions.MaxConcurrentCalls);
+        }
+
+        public async Task StartPumpAsync()
+        {
+            int retryCount = 0;
+            BrokeredMessage initialMessage = null;
+            while (true)
+            {
+                try
+                {
+                    initialMessage = await this.messageReceiver.ReceiveAsync().ConfigureAwait(false);
+                    MessagingEventSource.Log.MessageReceiverPumpInitialMessageReceived(this.messageReceiver.ClientId, initialMessage);
+                    break;
+                }
+                catch (Exception exception)
+                {
+                    retryCount++;
+                    MessagingEventSource.Log.MessageReceiverPumpInitialMessageReceiveException(this.messageReceiver.ClientId, retryCount, exception);
+
+                    if (retryCount == MaxInitialReceiveRetryCount ||
+                        !this.ShouldRetry(exception))
+                    {
+                        throw;
+                    }
+
+                    TimeSpan backOffTime = this.GetBackOffTime(exception);
+                    await Task.Delay(backOffTime, this.pumpCancellationToken).ConfigureAwait(false);
+                }
+            }
+
+            TaskExtensionHelper.Schedule(() => this.MessagePumpTask(initialMessage));
```

Comment:
This method tries to execute a method in a FireAndForget kind of mechanism, as in I want to execute this but do not wait for it to finish. Doing this just as is, without a corresponding await, will result in a CS4014, "because await is not called.. " warning.. Hence an extension to schedule this method to run in the context of another task.